extensions/spring-data-rest/deployment/src/main/java/io/quarkus/spring/data/rest/deployment/paging/PagingAndSortingMethodsImplementor.java

```diff
@@ -0,0 +1,182 @@
+package io.quarkus.spring.data.rest.deployment.paging;
+
+import static io.quarkus.gizmo.FieldDescriptor.of;
+import static io.quarkus.gizmo.MethodDescriptor.ofMethod;
+
+import java.util.Iterator;
+import java.util.List;
+
+import org.jboss.jandex.IndexView;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
+import org.springframework.data.repository.PagingAndSortingRepository;
+
+import io.quarkus.gizmo.AssignableResultHandle;
+import io.quarkus.gizmo.BranchResult;
+import io.quarkus.gizmo.BytecodeCreator;
+import io.quarkus.gizmo.ClassCreator;
+import io.quarkus.gizmo.MethodCreator;
+import io.quarkus.gizmo.MethodDescriptor;
+import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.panache.common.Page;
+import io.quarkus.rest.data.panache.deployment.Constants;
+import io.quarkus.spring.data.rest.deployment.crud.CrudMethodsImplementor;
+
+public class PagingAndSortingMethodsImplementor extends CrudMethodsImplementor {
+
+    public static final MethodDescriptor LIST_PAGED = ofMethod(PagingAndSortingRepository.class, "findAll",
+            org.springframework.data.domain.Page.class, Pageable.class);
+
+    private static final Class<?> PANACHE_PAGE = io.quarkus.panache.common.Page.class;
+
+    private static final Class<?> PANACHE_SORT = io.quarkus.panache.common.Sort.class;
+
+    private static final Class<?> PANACHE_COLUMN = io.quarkus.panache.common.Sort.Column.class;
+
+    private static final Class<?> PANACHE_DIRECTION = io.quarkus.panache.common.Sort.Direction.class;
+
+    public PagingAndSortingMethodsImplementor(IndexView index) {
+        super(index);
+    }
+
+    public void implementList(ClassCreator classCreator, String repositoryInterface) {
+        MethodCreator methodCreator = classCreator.getMethodCreator("list", List.class, Page.class,
+                io.quarkus.panache.common.Sort.class);
+
+        ResultHandle page = methodCreator.getMethodParam(0);
+        ResultHandle sort = methodCreator.getMethodParam(1);
+        ResultHandle pageable = toPageable(methodCreator, page, sort);
```

Comment:
Given we are taking the `sort` parameter from the REST URL, can we check it's not sensitive to SQL injections?

Reply:
Here page and sort are already `io.quarkus.panache.common.Page` and `io.quarkus.panache.common.Sort`. They're instantiated from the query parameters by `rest-data-panache` extension.
This extension (same as Hibernate and MongoDB Data REST alternatives) handle only data access logic. JAX-RS resource is handled in the same way for all of them by the `rest-data-panache`.

Reply:
I disagree with that.

Your extension makes it visible and applies absolutely no checks on the content coming from the URL. It's different when it's just Panache API, people can do their own checks.

So we need to check it's not sensitive to SQL injections otherwise you have a big security risk.

Reply:
I'm not saying that it shouldn't be checked. I'm saying that this is not the place to do it. This extension does not handle the JAX-RS logic. It only implements user defined interface that extends https://github.com/quarkusio/quarkus/blob/master/extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/RestDataResource.java.

`rest-data-panache` generates the JAX-RS resource and then calls `spring-data-rest` generated method for data access. So these two page and sort parameters are already instances of `io.quarkus.panache.common.Page` and `io.quarkus.panache.common.Sort`.

If we need to add query parameter validation we need to add it here https://github.com/quarkusio/quarkus/blob/master/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/ListMethodImplementor.java#L104. However, wouldn't Panache Query builder escape all the query parameters automatically?

Reply:
> I'm not saying that it shouldn't be checked. I'm saying that this is not the place to do it.

Well, I haven't really reviewed the other extension so I see it now for this one. I still think this should carefully be addressed.

> However, wouldn't Panache Query builder escape all the query parameters automatically?

It didn't a while ago. That's why I think we should check if it's safe or not.

Reply:
@FroMage maybe you know how unsafe it is to pass arbitrary parameters to the Panache Sort object? Are they somehow parsed down the line by Panache or Hibernate/Mongo? If this is unsafe, do you have a recommendation on what type of validation is needed?

In any case, I don't think that this issue should holdup this PR. If the validation has to be added, it will be added to the `rest-data-panache` extension.

Reply:
It's indeed unsafe because we end up generating an HQL query in `PanacheJpaUtil.toOrderBy` (not sure about Mongo but probably it is too).

Indeed the check really belongs in the REST layer as you say, especially wrt error handling. As to what you can check, here's what HQL allows:

```
orderByClause
// todo (6.0) : null precedence
	: ORDER BY sortSpecification (COMMA sortSpecification)*
	;

sortSpecification
	: expression collationSpecification? orderingSpecification?
	;

collationSpecification
	:	COLLATE collateName
	;

collateName
	:	dotIdentifierSequence
	;

orderingSpecification
	:	ASC
	|	DESC
	;
```

`expression` is particularly nasty because it's pretty much a free for all.

OTOH if your REST API specifies that the supported set of sort options is limited, then you can probably limit it to:

```
IDENTIFIER
	:	('a'..'z'|'A'..'Z'|'_'|'$'|'\u0080'..'\ufffe')('a'..'z'|'A'..'Z'|'_'|'$'|'0'..'9'|'\u0080'..'\ufffe')*
```

HTH :)

Reply:
BTW, all this is from the `panacheql` extension, you have the lexer and parser files there with all the info you need ;)

Reply:
Thanks @FroMage, this is very helpful. I'll add a query parameter validator to the `rest-data-extension` and will raise a new PR with it.

Reply:
I've created a separate pull request to add a sort parameter validation https://github.com/quarkusio/quarkus/pull/13514